highlight-wrapper/file-5.46/doc/file.1
.\" $File: file.man,v 1.151 2024/04/07 21:27:35 christos Exp $
.Dd April 7, 2024
.Dt FILE 1
.Os
.Sh NAME
.Nm file
.Nd determine file type
.Sh SYNOPSIS
.Nm
.Bk -words
.Op Fl bcdEhiklLNnprsSvzZ0
.Op Fl Fl apple
.Op Fl Fl exclude-quiet
.Op Fl Fl extension
.Op Fl Fl mime-encoding
.Op Fl Fl mime-type
.Op Fl e Ar testname
.Op Fl F Ar separator
.Op Fl f Ar namefile
.Op Fl m Ar magicfiles
.Op Fl P Ar name=value
.Ar
.Ek
.Nm
.Fl C
.Op Fl m Ar magicfiles
.Nm
.Op Fl Fl help
.Sh DESCRIPTION
This manual page documents version 5.46 of the
.Nm
command.
.Pp
.Nm
tests each argument in an attempt to classify it.
There are three sets of tests, performed in this order:
filesystem tests, magic tests, and language tests.
The
.Em first
test that succeeds causes the file type to be printed.
.Pp
The type printed will usually contain one of the words
.Em text
(the file contains only
printing characters and a few common control
characters and is probably safe to read on an
.Dv ASCII
terminal),
.Em executable
(the file contains the result of compiling a program
in a form understandable to some
.Tn UNIX
kernel or another),
or
.Em data
meaning anything else (data is usually
.Dq binary
or non-printable).
Exceptions are well-known file formats (core files, tar archives)
that are known to contain binary data.
When modifying magic files or the program itself, make sure to
.Em preserve these keywords .
Users depend on knowing that all the readable files in a directory
have the word
.Dq text
printed.
Don't do as Berkeley did and change
.Dq shell commands text
to
.Dq shell script .
.Pp
The filesystem tests are based on examining the return from a
.Xr stat 2
system call.
The program checks to see if the file is empty,
or if it's some sort of special file.
Any known file types appropriate to the system you are running on
(sockets, symbolic links, or named pipes (FIFOs) on those systems that
implement them)
are intuited if they are defined in the system header file
.In sys/stat.h .
.Pp
The magic tests are used to check for files with data in
particular fixed formats.
The canonical example of this is a binary executable (compiled program)
.Dv a.out
file, whose format is defined in
.In elf.h ,
.In a.out.h
and possibly
.In exec.h
in the standard include directory.
These files have a
.Dq magic number
stored in a particular place
near the beginning of the file that tells the
.Tn UNIX
operating system
that the file is a binary executable, and which of several types thereof.
The concept of a
.Dq magic number
has been applied by extension to data files.
Any file with some invariant identifier at a small fixed
offset into the file can usually be described in this way.
The information identifying these files is read from the compiled
magic file
.Pa /usr/local/share/misc/magic.mgc ,
or the files in the directory
.Pa /usr/local/share/misc/magic
if the compiled file does not exist.
In addition, if
.Pa $HOME/.magic.mgc
or
.Pa $HOME/.magic
exists, it will be used in preference to the system magic files.
.Pp
If a file does not match any of the entries in the magic file,
it is examined to see if it seems to be a text file.
ASCII, ISO-8859-x, non-ISO 8-bit extended-ASCII character sets
(such as those used on Macintosh and IBM PC systems),
UTF-8-encoded Unicode, UTF-16-encoded Unicode, and EBCDIC
character sets can be distinguished by the different
ranges and sequences of bytes that constitute printable text
in each set.
If a file passes any of these tests, its character set is reported.
ASCII, ISO-8859-x, UTF-8, and extended-ASCII files are identified
as
.Dq text
because they will be mostly readable on nearly any terminal;
UTF-16 and EBCDIC are only
.Dq character data
because, while
they contain text, it is text that will require translation
before it can be read.
In addition,
.Nm
will attempt to determine other characteristics of text-type files.
If the lines of a file are terminated by CR, CRLF, or NEL, instead
of the Unix-standard LF, this will be reported.
Files that contain embedded escape sequences or overstriking
will also be identified.
.Pp
Once
.Nm
has determined the character set used in a text-type file,
it will
attempt to determine in what language the file is written.
The language tests look for particular strings (cf.
.In names.h )
that can appear anywhere in the first few blocks of a file.
For example, the keyword
.Em .br
indicates that the file is most likely a
.Xr troff 1
input file, just as the keyword
.Em struct
indicates a C program.
These tests are less reliable than the previous
two groups, so they are performed last.
The language test routines also test for some miscellany
(such as
.Xr tar 1
archives, JSON files).
.Pp
Any file that cannot be identified as having been written
in any of the character sets listed above is simply said to be
.Dq data .
.Sh OPTIONS
.Bl -tag -width indent
.It Fl Fl apple
Causes the
.Nm
command to output the file type and creator code as
used by older MacOS versions.
The code consists of eight letters,
the first describing the file type, the latter the creator.
This option works properly only for file formats that have the
apple-style output defined.
.It Fl b , Fl Fl brief
Do not prepend filenames to output lines (brief mode).
.It Fl C , Fl Fl compile
Write a
.Pa magic.mgc
output file that contains a pre-parsed version of the magic file or directory.
.It Fl c , Fl Fl checking-printout
Cause a checking printout of the parsed form of the magic file.
This is usually used in conjunction with the
.Fl m
option to debug a new magic file before installing it.
.It Fl d
Prints internal debugging information to stderr.
.It Fl E
On filesystem errors (file not found etc), instead of handling the error
as regular output as POSIX mandates and keep going, issue an error message
and exit.
.It Fl e , Fl Fl exclude Ar testname
Exclude the test named in
.Ar testname
from the list of tests made to determine the file type.
Valid test names are:
.Bl -tag -width compress
.It apptype
.Dv EMX
application type (only on EMX).
.It ascii
Various types of text files (this test will try to guess the text
encoding, irrespective of the setting of the
.Sq encoding
option).
.It encoding
Different text encodings for soft magic tests.
.It tokens
Ignored for backwards compatibility.
.It cdf
Prints details of Compound Document Files.
.It compress
Checks for, and looks inside, compressed files.
.It csv
Checks Comma Separated Value files.
.It elf
Prints ELF file details, provided soft magic tests are enabled and the
elf magic is found.
.It json
Examines JSON (RFC-7159) files by parsing them for compliance.
.It soft
Consults magic files.
.It simh
Examines SIMH tape files.
.It tar
Examines tar files by verifying the checksum of the 512 byte tar header.
Excluding this test can provide more detailed content description by using
the soft magic method.
.It text
A synonym for
.Sq ascii .
.El
.It Fl Fl exclude-quiet
Like
.Fl Fl exclude
but ignore tests that
.Nm
does not know about.
This is intended for compatibility with older versions of
.Nm .
.It Fl Fl extension
Print a slash-separated list of valid extensions for the file type found.
.It Fl F , Fl Fl separator Ar separator
Use the specified string as the separator between the filename and the
file result returned.
Defaults to
.Sq \&: .
.It Fl f , Fl Fl files-from Ar namefile
Read the names of the files to be examined from
.Ar namefile
(one per line)
before the argument list.
Either
.Ar namefile
or at least one filename argument must be present;
to test the standard input, use
.Sq -
as a filename argument.
Please note that
.Ar namefile
is unwrapped and the enclosed filenames are processed when this option is

CHANGELOG.md
# Changelog


### 1.0.23 (49)
Bugfix:
- Unable to save the settings.


### 1.0.22 (47)
New Features:
- Support for MDX files (`.mdx`, _without_ JSX rendering).
- Support for Cursor Rulers files (`.mdc`).
- Highlight updated to 4.16.
- Lua updated to 5.4.7.
- File updated to 5.46.
- Enry updated to 2.9.2.
- Support for UTI `io.typora.markdown`.

Bugfix:
- Rewrite settings management to bypass warnings about accessing settings from other applications. **You will need to reset any preferences you have changed.**
- Unused code has been removed (including the ability to choose a theme for code formatting, now to change it you will need to edit the css file)


### 1.0.21 (46)
New Features:
- Warning about this is not a Markdown editor app.
- Alert when the app is closed and the current markdown file was changed.
- Experimental Shortcut Action (require macOS 15.2). 

Bugfix:
- Temporary fix for $ recognition without mathematical expression.


### 1.0.20 (45)
New Features:
- Support for some extended syntax:
  - highlight text with `==`
  - subscript (`~`) and superscript (`^`).


### 1.0.19 (44)
Bugfix:
- Fixed saving about flag settings.


### 1.0.18 (43)
New Features:
- Support for `com.nutstore.down` UTI.
- cmark-gfm updated to version 0.29.0.gfm.13.
- pcre2 updated to version 10.43.
- Enry updated to version 2.8.8.
- Settings to show about app info on the footer preview.
- Link to buy me a :coffee: (:heart:)

Bugfix:
- Exported html head title filled with the filename.


### 1.0.17 (42)
New Features:
- Improved Math (Latex) support (thanks to @mccann):
    - Prevent Markdown processing inside $'s
    - Allow multi-line Latex, like that supported by:
        - jupyter notebook
        - vscode
        - obsidian
        - probably many more - only tested the above.
    - Now prevents arbitrary html getting passed through.


### 1.0.16 (41)
New features:
- Support mathematical expressions with the MathJax engine.
- Update libraries:
    - cmark-gfm 0.29.0.gfm.11
    - highlight 4.5
    - lua 5.4.4
    - libmagic 5.44
    - go-enry v2.8.4
    - boost.
- Settings UI redesigned.
  
Bugfix:
- Alt attribute for emoji rendered as image.
 

### 1.0.15 (40)
New features:
- The markdown file is automatically reloaded when it is edited outside of the app.  
- Preliminary support for Quarto files (`.qmd`).


### 1.0.14 (39)
Bugfix:
- Allows you to use `...` to end the `yaml` header block. 


### 1.0.13 (38)
Bugfix:
- Fixed the Sparkle integration bug. **If you have installed version 1.0.11 or 1.0.12 you may need to [re-download the updated app from the web](https://github.com/sbarex/QLMarkdown/releases/download/1.0.13/QLMarkdown.zip).** 


### 1.0.12 (37)
Bugfix:
- Better procedure to install the command line tool.
- Fixed the bug that prevented the processing of html raw images when they are not inserted in a html block element. 


### 1.0.11 (36)
New features:
- Support for opening markdown files (by dragging the file onto the app icon)
- Support for exporting the markdown code.
- Sparkle updated to release 2.0.0. 

Bugfix:
- Fix for heads with dash.
- Implemented missing behavior for the color scheme editor.
- Fix for installation of the command line tool.


### 1.0.10 (35)
New features:
- Better performance for heads extension.
- Better performance for inline images on raw html fragments.
- Option for automatic saving of settings changes.
- GUI optimization.

Bugfix:
- Fixed settings save.


### 1.0.9 (34)
New features:
- Experimental support for defining the size of the Quick Look window.
- Magic `file` updated to release 5.41.

Bugfix:
- Fixed cross compile compilation of `file` library.


### 1.0.8 (33)
Bugfix:
- Fixed bug for undefined source code theme.
- Fixed bug on settings sync delay in the Quick Look preview on macOS Monterey.


### 1.0.7 (32)

highlight-wrapper/file-5.44/src/file.c
/*
 * Copyright (c) Ian F. Darwin 1986-1995.
 * Software written by Ian F. Darwin and others;
 * maintained 1995-present by Christos Zoulas and others.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice immediately at the beginning of the file, without modification,
 *    this list of conditions, and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
/*
 * file - find type of a file or files - main program.
 */

#include "file.h"

#ifndef	lint
FILE_RCSID("@(#)$File: file.c,v 1.213 2022/12/26 17:31:14 christos Exp $")
#endif	/* lint */

#include "magic.h"

#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#ifdef RESTORE_TIME
# if (__COHERENT__ >= 0x420)
#  include <sys/utime.h>
# else
#  ifdef USE_UTIMES
#   include <sys/time.h>
#  else
#   include <utime.h>
#  endif
# endif
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>	/* for read() */
#endif
#ifdef HAVE_WCHAR_H
#include <wchar.h>
#endif
#ifdef HAVE_WCTYPE_H
#include <wctype.h>
#endif
#if defined(HAVE_WCHAR_H) && defined(HAVE_MBRTOWC) && defined(HAVE_WCWIDTH) && \
   defined(HAVE_WCTYPE_H)
#define FILE_WIDE_SUPPORT
#else
#include <ctype.h>
#endif

#if defined(HAVE_GETOPT_H) && defined(HAVE_STRUCT_OPTION)
# include <getopt.h>
# ifndef HAVE_GETOPT_LONG
int getopt_long(int, char * const *, const char *,
    const struct option *, int *);
# endif
# else
#  include "mygetopt.h"
#endif

#ifdef S_IFLNK
# define IFLNK_h "h"
# define IFLNK_L "L"
#else
# define IFLNK_h ""
# define IFLNK_L ""
#endif

#define FILE_FLAGS	"bcCdE" IFLNK_h "ik" IFLNK_L "lNnprsSvzZ0"
#define OPTSTRING	"bcCde:Ef:F:hiklLm:nNpP:rsSvzZ0"

# define USAGE  \
    "Usage: %s [-" FILE_FLAGS "] [--apple] [--extension] [--mime-encoding]\n" \
    "            [--mime-type] [-e <testname>] [-F <separator>] " \
    " [-f <namefile>]\n" \
    "            [-m <magicfiles>] [-P <parameter=value>] [--exclude-quiet]\n" \
    "            <file> ...\n" \
    "       %s -C [-m <magicfiles>]\n" \
    "       %s [--help]\n"

file_private int 		/* Global command-line options 		*/
	bflag = 0,	/* brief output format	 		*/
	nopad = 0,	/* Don't pad output			*/
	nobuffer = 0,   /* Do not buffer stdout 		*/
	nulsep = 0;	/* Append '\0' to the separator		*/

file_private const char *separator = ":";	/* Default field separator	*/
file_private const struct option long_options[] = {
#define OPT_HELP		1
#define OPT_APPLE		2
#define OPT_EXTENSIONS		3
#define OPT_MIME_TYPE		4
#define OPT_MIME_ENCODING	5
#define OPT_EXCLUDE_QUIET	6
#define OPT(shortname, longname, opt, def, doc)		\
    {longname, opt, NULL, shortname},
#define OPT_LONGONLY(longname, opt, def, doc, id)	\
    {longname, opt, NULL, id},
#include "file_opts.h"
#undef OPT
#undef OPT_LONGONLY
    {0, 0, NULL, 0}
    };

file_private const struct {
	const char *name;
	int value;
} nv[] = {
	{ "apptype",	MAGIC_NO_CHECK_APPTYPE },
	{ "ascii",	MAGIC_NO_CHECK_ASCII },
	{ "cdf",	MAGIC_NO_CHECK_CDF },
	{ "compress",	MAGIC_NO_CHECK_COMPRESS },
	{ "csv",	MAGIC_NO_CHECK_CSV },
	{ "elf",	MAGIC_NO_CHECK_ELF },
	{ "encoding",	MAGIC_NO_CHECK_ENCODING },
	{ "soft",	MAGIC_NO_CHECK_SOFT },
	{ "tar",	MAGIC_NO_CHECK_TAR },
	{ "json",	MAGIC_NO_CHECK_JSON },
	{ "text",	MAGIC_NO_CHECK_TEXT },	/* synonym for ascii */
	{ "tokens",	MAGIC_NO_CHECK_TOKENS }, /* OBSOLETE: ignored for backwards compatibility */
};

file_private struct {
	const char *name;
	size_t value;
	size_t def;
	const char *desc;
	int tag;
	int set;
} pm[] = {
	{ "bytes", 0, FILE_BYTES_MAX, "max bytes to look inside file",
	    MAGIC_PARAM_BYTES_MAX, 0 },
	{ "elf_notes", 0, FILE_ELF_NOTES_MAX, "max ELF notes processed",
	    MAGIC_PARAM_ELF_NOTES_MAX, 0 },
	{ "elf_phnum", 0, FILE_ELF_PHNUM_MAX, "max ELF prog sections processed",
	    MAGIC_PARAM_ELF_PHNUM_MAX, 0 },
	{ "elf_shnum", 0, FILE_ELF_SHNUM_MAX, "max ELF sections processed",
	    MAGIC_PARAM_ELF_SHNUM_MAX, 0 },
	{ "encoding", 0, FILE_ENCODING_MAX, "max bytes to scan for encoding",
	    MAGIC_PARAM_ENCODING_MAX, 0 },
	{ "indir", 0, FILE_INDIR_MAX, "recursion limit for indirection",
	    MAGIC_PARAM_INDIR_MAX, 0 },
	{ "name", 0, FILE_NAME_MAX, "use limit for name/use magic",
	    MAGIC_PARAM_NAME_MAX, 0 },
	{ "regex", 0, FILE_REGEX_MAX, "length limit for REGEX searches",
	    MAGIC_PARAM_REGEX_MAX, 0 },
};

file_private int posixly;

#ifdef __dead
__dead
#endif
file_private void usage(void);
file_private void docprint(const char *, int);
#ifdef __dead
__dead
#endif
file_private void help(void);

file_private int unwrap(struct magic_set *, const char *);
file_private int process(struct magic_set *ms, const char *, int);
file_private struct magic_set *load(const char *, int);
file_private void setparam(const char *);
file_private void applyparam(magic_t);


/*
 * main - parse arguments and handle options
 */
int
main(int argc, char *argv[])
{
	int c;
	size_t i, j, wid, nw;
	int action = 0, didsomefiles = 0, errflg = 0;
	int flags = 0, e = 0;
#ifdef HAVE_LIBSECCOMP
	int sandbox = 1;
#endif
	struct magic_set *magic = NULL;
	int longindex;
	const char *magicfile = NULL;		/* where the magic is	*/
	char *progname;

	/* makes islower etc work for other langs */
	(void)setlocale(LC_CTYPE, "");

#ifdef __EMX__
	/* sh-like wildcard expansion! Shouldn't hurt at least ... */
	_wildcard(&argc, &argv);
#endif

	if ((progname = strrchr(argv[0], '/')) != NULL)
		progname++;
	else
		progname = argv[0];

	file_setprogname(progname);


#ifdef S_IFLNK
	posixly = getenv("POSIXLY_CORRECT") != NULL;
	flags |=  posixly ? MAGIC_SYMLINK : 0;
#endif
	while ((c = getopt_long(argc, argv, OPTSTRING, long_options,
	    &longindex)) != -1)
		switch (c) {
		case OPT_HELP:
			help();
			break;
		case OPT_APPLE:
			flags |= MAGIC_APPLE;
			break;
		case OPT_EXTENSIONS:
			flags |= MAGIC_EXTENSION;
			break;
		case OPT_MIME_TYPE:
			flags |= MAGIC_MIME_TYPE;
			break;
		case OPT_MIME_ENCODING:
			flags |= MAGIC_MIME_ENCODING;
			break;
		case '0':
			nulsep++;
			break;
		case 'b':
			bflag++;
			break;
		case 'c':
			action = FILE_CHECK;
			break;
		case 'C':
			action = FILE_COMPILE;
			break;
		case 'd':
			flags |= MAGIC_DEBUG|MAGIC_CHECK;
			break;
		case 'E':
			flags |= MAGIC_ERROR;
			break;
		case 'e':
		case OPT_EXCLUDE_QUIET:
			for (i = 0; i < __arraycount(nv); i++)
				if (strcmp(nv[i].name, optarg) == 0)
					break;

			if (i == __arraycount(nv)) {
				if (c != OPT_EXCLUDE_QUIET)
					errflg++;
			} else
				flags |= nv[i].value;
			break;

		case 'f':
			if(action)
				usage();
			if (magic == NULL)
				if ((magic = load(magicfile, flags)) == NULL)
					return 1;
			applyparam(magic);
			e |= unwrap(magic, optarg);
			++didsomefiles;
			break;
		case 'F':
			separator = optarg;
			break;
		case 'i':
			flags |= MAGIC_MIME;
			break;
		case 'k':
			flags |= MAGIC_CONTINUE;
			break;
		case 'l':
			action = FILE_LIST;
			break;
		case 'm':
			magicfile = optarg;
			break;
		case 'n':
			++nobuffer;
			break;
		case 'N':
			++nopad;
			break;
#if defined(HAVE_UTIME) || defined(HAVE_UTIMES)
		case 'p':
			flags |= MAGIC_PRESERVE_ATIME;
			break;
#endif
		case 'P':
			setparam(optarg);
			break;
		case 'r':
			flags |= MAGIC_RAW;
			break;
		case 's':
			flags |= MAGIC_DEVICES;
			break;
		case 'S':
#ifdef HAVE_LIBSECCOMP
			sandbox = 0;
#endif
			break;
		case 'v':
			if (magicfile == NULL)
				magicfile = magic_getpath(magicfile, action);
			(void)fprintf(stdout, "%s-%s\n", file_getprogname(),
			    VERSION);
			(void)fprintf(stdout, "magic file from %s\n",
			    magicfile);
#ifdef HAVE_LIBSECCOMP
			(void)fprintf(stdout, "seccomp support included\n");
#endif
			return 0;
		case 'z':
			flags |= MAGIC_COMPRESS;
			break;

		case 'Z':
			flags |= MAGIC_COMPRESS|MAGIC_COMPRESS_TRANSP;
			break;
#ifdef S_IFLNK
		case 'L':
			flags |= MAGIC_SYMLINK;
			break;
		case 'h':
			flags &= ~MAGIC_SYMLINK;
			break;
#endif
		case '?':
		default:
			errflg++;
			break;
		}

	if (errflg) {
		usage();
	}
	if (e)
		return e;

#ifdef HAVE_LIBSECCOMP
#if 0
	if (sandbox && enable_sandbox_basic() == -1)
#else
	if (sandbox && enable_sandbox_full() == -1)
#endif
		file_err(EXIT_FAILURE, "SECCOMP initialisation failed");
	if (sandbox)
		flags |= MAGIC_NO_COMPRESS_FORK;
#endif /* HAVE_LIBSECCOMP */

	if (MAGIC_VERSION != magic_version())
		file_warnx("Compiled magic version [%d] "
		    "does not match with shared library magic version [%d]\n",
		    MAGIC_VERSION, magic_version());

	switch(action) {
	case FILE_CHECK:
	case FILE_COMPILE:
	case FILE_LIST:
		/*
		 * Don't try to check/compile ~/.magic unless we explicitly
		 * ask for it.
		 */
		magic = magic_open(flags|MAGIC_CHECK);
		if (magic == NULL) {
			file_warn("Can't create magic");
			return 1;
		}


		switch(action) {
		case FILE_CHECK:
			c = magic_check(magic, magicfile);
			break;
		case FILE_COMPILE:
			c = magic_compile(magic, magicfile);
			break;
		case FILE_LIST:
			c = magic_list(magic, magicfile);
			break;
		default:
			abort();
		}
		if (c == -1) {
			file_warnx("%s", magic_error(magic));
			e = 1;
			goto out;
		}
		goto out;
	default:
		if (magic == NULL)
			if ((magic = load(magicfile, flags)) == NULL)
				return 1;
		applyparam(magic);
	}

	if (optind == argc) {
		if (!didsomefiles)
			usage();
		goto out;
	}

	for (wid = 0, j = CAST(size_t, optind); j < CAST(size_t, argc);
	    j++) {
		nw = file_mbswidth(magic, argv[j]);
		if (nw > wid)
			wid = nw;
	}

	/*
	 * If bflag is only set twice, set it depending on
	 * number of files [this is undocumented, and subject to change]
	 */
	if (bflag == 2) {
		bflag = optind >= argc - 1;
	}
	for (; optind < argc; optind++)
		e |= process(magic, argv[optind], wid);

out:
	if (!nobuffer)
		e |= fflush(stdout) != 0;

	if (magic)
		magic_close(magic);
	return e;
}

file_private void
applyparam(magic_t magic)
{
	size_t i;

	for (i = 0; i < __arraycount(pm); i++) {
		if (!pm[i].set)
			continue;
		if (magic_setparam(magic, pm[i].tag, &pm[i].value) == -1)
			file_err(EXIT_FAILURE, "Can't set %s", pm[i].name);
	}
}

file_private void
setparam(const char *p)
{
	size_t i;
	char *s;

	if ((s = CCAST(char *, strchr(p, '='))) == NULL)
		goto badparm;

	for (i = 0; i < __arraycount(pm); i++) {
		if (strncmp(p, pm[i].name, s - p) != 0)
			continue;
		pm[i].value = atoi(s + 1);
		pm[i].set = 1;
		return;
	}
badparm:
	file_errx(EXIT_FAILURE, "Unknown param %s", p);
}

file_private struct magic_set *
/*ARGSUSED*/
load(const char *magicfile, int flags)
{
	struct magic_set *magic = magic_open(flags);
	const char *e;

	if (magic == NULL) {
		file_warn("Can't create magic");
		return NULL;
	}
	if (magic_load(magic, magicfile) == -1) {
		file_warn("%s", magic_error(magic));
		magic_close(magic);
		return NULL;
	}
	if ((e = magic_error(magic)) != NULL)
		file_warn("%s", e);
	return magic;
}

/*
 * unwrap -- read a file of filenames, do each one.
 */
file_private int
unwrap(struct magic_set *ms, const char *fn)
{
	FILE *f;
	ssize_t len;
	char *line = NULL;
	size_t llen = 0;
	int wid = 0, cwid;
	int e = 0;
	size_t fi = 0, fimax = 0;
	char **flist = NULL;

	if (strcmp("-", fn) == 0)
		f = stdin;
	else {
		if ((f = fopen(fn, "r")) == NULL) {
			file_warn("Cannot open `%s'", fn);
			return 1;
		}
	}

	while ((len = getline(&line, &llen, f)) > 0) {
		if (line[len - 1] == '\n')
			line[len - 1] = '\0';
		cwid = file_mbswidth(ms, line);
		if (nobuffer) {
			e |= process(ms, line, cwid);
			free(line);
			line = NULL;
			llen = 0;
			continue;
		}
		if (cwid > wid)
			wid = cwid;
		if (fi >= fimax) {
			fimax += 100;
			char **nf = CAST(char **,
			    realloc(flist, fimax * sizeof(*flist)));
			if (nf == NULL) {
				file_err(EXIT_FAILURE,
				    "Cannot allocate memory for file list");
			}
			flist = nf;
		}
		flist[fi++] = line;
		line = NULL;
		llen = 0;
	}

	if (!nobuffer) {
		fimax = fi;
		for (fi = 0; fi < fimax; fi++) {
			e |= process(ms, flist[fi], wid);
			free(flist[fi]);
		}
	}
	free(flist);

	if (f != stdin)
		(void)fclose(f);
	return e;
}

file_private void
file_octal(unsigned char c)
{
	(void)putc('\\', stdout);
	(void)putc(((c >> 6) & 7) + '0', stdout);
	(void)putc(((c >> 3) & 7) + '0', stdout);
	(void)putc(((c >> 0) & 7) + '0', stdout);
}

file_private void
fname_print(const char *inname)
{
	size_t n = strlen(inname);
#ifdef FILE_WIDE_SUPPORT
	mbstate_t state;
	wchar_t nextchar;
	size_t bytesconsumed;


	(void)memset(&state, 0, sizeof(state));
	while (n > 0) {
		bytesconsumed = mbrtowc(&nextchar, inname, n, &state);
		if (bytesconsumed == CAST(size_t, -1) ||
		    bytesconsumed == CAST(size_t, -2))  {
			nextchar = *inname++;
			n--;
			(void)memset(&state, 0, sizeof(state));
			file_octal(CAST(unsigned char, nextchar));
			continue;
		}
		inname += bytesconsumed;
		n -= bytesconsumed;
		if (iswprint(nextchar)) {
			printf("%lc", (wint_t)nextchar);
			continue;
		}
		/* XXX: What if it is > 255? */
		file_octal(CAST(unsigned char, nextchar));
	}
#else
	size_t i;
	for (i = 0; i < n; i++) {
		unsigned char c = CAST(unsigned char, inname[i]);
		if (isprint(c)) {
			(void)putc(c, stdout);
			continue;
		}
		file_octal(c);
	}
#endif
}

/*
 * Called for each input file on the command line (or in a list of files)
 */
file_private int
process(struct magic_set *ms, const char *inname, int wid)
{
	const char *type, c = nulsep > 1 ? '\0' : '\n';
	int std_in = strcmp(inname, "-") == 0;
	int haderror = 0;

	if (wid > 0 && !bflag) {
		const char *pname = std_in ? "/dev/stdin" : inname;
		if ((ms->flags & MAGIC_RAW) == 0)
			fname_print(pname);
		else
			(void)printf("%s", pname);
		if (nulsep)
			(void)putc('\0', stdout);
		if (nulsep < 2) {
			(void)printf("%s", separator);
			(void)printf("%*s ", CAST(int, nopad ? 0
			    : (wid - file_mbswidth(ms, inname))), "");
		}
	}

	type = magic_file(ms, std_in ? NULL : inname);

	if (type == NULL) {
		haderror |= printf("ERROR: %s%c", magic_error(ms), c);
	} else {
		haderror |= printf("%s%c", type, c) < 0;
	}
	if (nobuffer)
		haderror |= fflush(stdout) != 0;
	return haderror || type == NULL;
}

file_protected size_t
file_mbswidth(struct magic_set *ms, const char *s)
{
	size_t width = 0;
#ifdef FILE_WIDE_SUPPORT
	size_t bytesconsumed, n;
	mbstate_t state;
	wchar_t nextchar;

	(void)memset(&state, 0, sizeof(state));
	n = strlen(s);

	while (n > 0) {
		bytesconsumed = mbrtowc(&nextchar, s, n, &state);
		if (bytesconsumed == CAST(size_t, -1) ||
		    bytesconsumed == CAST(size_t, -2)) {
			nextchar = *s;
			bytesconsumed = 1;
			(void)memset(&state, 0, sizeof(state));
			width += 4;
		} else {
			int w = wcwidth(nextchar);
			width += ((ms->flags & MAGIC_RAW) != 0
			    || iswprint(nextchar)) ? (w > 0 ? w : 1) : 4;
		}

		s += bytesconsumed, n -= bytesconsumed;
	}
#else
	for (; *s; s++) {
		width += (ms->flags & MAGIC_RAW) != 0
		    || isprint(CAST(unsigned char, *s)) ? 1 : 4;
	}
#endif
	return width;
}

file_private void
usage(void)
{
	const char *pn = file_getprogname();
	(void)fprintf(stderr, USAGE, pn, pn, pn);
	exit(EXIT_FAILURE);
}

file_private void
defprint(int def)
{
	if (!def)
		return;
	if (((def & 1) && posixly) || ((def & 2) && !posixly))
		(void)fprintf(stdout, " (default)");
	(void)putc('\n', stdout);
}

file_private void
docprint(const char *opts, int def)
{
	size_t i;
	int comma, pad;
	char *sp, *p;

	p = CCAST(char *, strchr(opts, '%'));
	if (p == NULL) {
		(void)fprintf(stdout, "%s", opts);
		defprint(def);
		return;
	}

	for (sp = p - 1; sp > opts && *sp == ' '; sp--)
		continue;

	(void)printf("%.*s", CAST(int, p - opts), opts);
	pad = (int)CAST(int, p - sp - 1);

	switch (*++p) {
	case 'e':
		comma = 0;
		for (i = 0; i < __arraycount(nv); i++) {
			(void)printf("%s%s", comma++ ? ", " : "", nv[i].name);
			if (i && i % 5 == 0 && i != __arraycount(nv) - 1) {
				(void)printf(",\n%*s", pad, "");
				comma = 0;
			}
		}
		break;
	case 'P':
		for (i = 0; i < __arraycount(pm); i++) {
			(void)printf("%9s %7zu %s", pm[i].name, pm[i].def,
			    pm[i].desc);
			if (i != __arraycount(pm) - 1)
				(void)printf("\n%*s", pad, "");
		}
		break;
	default:
		file_errx(EXIT_FAILURE, "Unknown escape `%c' in long options",
		   *p);
		break;
	}
	(void)printf("%s", opts + (p - opts) + 1);

}

file_private void
help(void)
{
	(void)fputs(
"Usage: file [OPTION...] [FILE...]\n"
"Determine type of FILEs.\n"
"\n", stdout);
#define OPT(shortname, longname, opt, def, doc)      \
	(void)printf("  -%c, --" longname, shortname), \
	docprint(doc, def);
#define OPT_LONGONLY(longname, opt, def, doc, id)    	\
	(void)printf("      --" longname),	\
	docprint(doc, def);
#include "file_opts.h"
#undef OPT
#undef OPT_LONGONLY
	(void)printf("\nReport bugs to https://bugs.astron.com/\n");
	exit(EXIT_SUCCESS);
}

file_private const char *file_progname;

file_protected void
file_setprogname(const char *progname)
{
	file_progname = progname;
}

file_protected const char *
file_getprogname(void)
{
	return file_progname;
}

file_protected void
file_err(int e, const char *fmt, ...)
{
	va_list ap;
	int se = errno;

	va_start(ap, fmt);
	(void)fprintf(stderr, "%s: ", file_progname);
	(void)vfprintf(stderr, fmt, ap);
	va_end(ap);
	if (se)
		(void)fprintf(stderr, " (%s)\n", strerror(se));
	else
		fputc('\n', stderr);
	exit(e);
}

file_protected void
file_errx(int e, const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	(void)fprintf(stderr, "%s: ", file_progname);
	(void)vfprintf(stderr, fmt, ap);
	va_end(ap);
	(void)fprintf(stderr, "\n");
	exit(e);
}

file_protected void
file_warn(const char *fmt, ...)
{
	va_list ap;
	int se = errno;

	va_start(ap, fmt);
	(void)fprintf(stderr, "%s: ", file_progname);
	(void)vfprintf(stderr, fmt, ap);
	va_end(ap);
	if (se)
		(void)fprintf(stderr, " (%s)\n", strerror(se));
	else
		fputc('\n', stderr);
	errno = se;
}

file_protected void
file_warnx(const char *fmt, ...)
{
	va_list ap;
	int se = errno;

	va_start(ap, fmt);
	(void)fprintf(stderr, "%s: ", file_progname);
	(void)vfprintf(stderr, fmt, ap);
	va_end(ap);
	(void)fprintf(stderr, "\n");
	errno = se;
}
